Staking contract  smiley (#243)
near/NEPs: archive/0000-staking-contract.md

- Proposal Name: staking-contract
- Start Date: 2020-01-12
- NEP PR: [nearprotocol/neps#0000](https://github.com/nearprotocol/neps/pull/0000)
- Issue(s): link to relevant issues in relevant repos (not required).

# Summary
[summary]: #summary

Provides standard and reference internal design for staking contract with delegation.

# Motivation

> [motivation]: #motivation

Reviewer: ```suggestion
```

Reviewer: this is not needed. Let's remove it in other places as well.

> 

Reviewer: let's start with explanation why we need this standard. What does it bring to the community and why it is used.

In NEAR contracts and accounts are the same object. This allows to have a contract that stakes NEAR with it's balance.
This is useful to implement delegation, custom reward dispersion, liquid staking and various derivatives on stake and more.

# Guide-level explanation
[guide-level-explanation]: #guide-level-explanation

The simplest example of staking contract would be a delegation contract.

Let's define actors:
- The staking pool contract account `staking-pool`. A key-less account with the contract that pools funds.
- The owner of the staking contract `owner`. Owner runs the validator node on behalf of the staking pool account.
- A delegator `user1`. The account who wants to stake their fund to the pool.

> The owner can setup such contract and validate on behalf of this contract in their node.
> Any other user can send their tokens to the contract, which will be pooled together and increase the total stake.
> These users would accrue rewards (subtracted fees set by the owner).
> Then they can unstake and withdraw their balance after some unlocking period.

Reviewer: ```suggestion
Usually, owner owns the contract and is the validator. Delegators will stake through the contract. The stake will be pooled and if enough stake is collected, it will allow the validator to be an active validator and consequently receive staking rewards. The default implementation is fully permission-less and automates the staking reward distribution as well as the validator fees.
Delegators can unstake and withdraw their balance after the unlocking period.
```

More complex example can also issue token for deposited user's stake. This stake is the right to withdraw underlaying balance and rewards. This provides staking liquidity and allows to trade staked tokens.

# Reference-level explanation
[reference-level-explanation]: #reference-level-explanation

> ## Staking contract guarantees:

Reviewer: ```suggestion
## Requirements
```

- The contract can't lose or lock tokens of users.

> - If a user deposited X, the user should be able to withdraw at least X.

Reviewer: What happens if a validator will get slashed... I think it's possible that validator will receive less.

- If a user successfully staked X, the user can unstake at least X.

> - The contract should lock unstaked funds for longer than 4 epochs.

Reviewer: let's add more clarification:
* why 4 epochs? Maybe let's link to the staking documentation
* add to the requirements about fees and risks

/// Stakes the given amount from the inner account of the predecessor.
/// The inner account should have enough unstaked balance.
pub fn stake(&mut self, amount: U128);

/// Unstakes the given amount from the inner account of the predecessor.
/// The inner account should have enough staked balance.
/// The new total unstaked balance will be available for withdrawal in 4 epochs.
pub fn unstake(&mut self, amount: U128);

/****************/
/* View methods */
/****************/

/// Returns the unstaked balance of the given account.
pub fn get_account_unstaked_balance(&self, account_id: AccountId) -> U128;

/// Returns the staked balance of the given account.
pub fn get_account_staked_balance(&self, account_id: AccountId) -> U128;

/// Returns the total balance of the given account (including staked and unstaked balances).
pub fn get_account_total_balance(&self, account_id: AccountId) -> U128;

/// Returns `true` if the given account can withdraw unstaked tokens in the current epoch.
pub fn is_account_unstaked_balance_available(&self, account_id: AccountId) -> bool;

/// Returns the total staking balance.
pub fn get_total_staked_balance(&self) -> U128;
```

> ## User path

Reviewer: ```suggestion
## User flows
```
